### PR TITLE
adding mermaidjs shortcode with example

### PR DESCRIPTION
**Pull request by jimangel**
Enabling visuals for https://mermaid-js.github.io/mermaid/ via shortcode. This was discussed at the weekly docs meeting on 4/14/2020.

I've included an example markdown doc too.

### CODE REVIEW

**Comment by jimangel** [comment]
Related: https://github.com/kubernetes/website/issues/20293

**Comment by jimangel** [comment]
Preview: https://deploy-preview-20434--kubernetes-io-master-staging.netlify.app/docs/test/#visualizations-with-mermaid

File: layouts/partials/css.html [hunk 1]
```diff
@@ -19,6 +19,9 @@
 <link rel="stylesheet" href="{{ "css/jquery-ui.min.css" | relURL }}">
 <link rel="stylesheet" href="{{ "css/callouts.css" | relURL }}">
 <link rel="stylesheet" href="{{ "css/custom-jekyll/tags.css" | relURL }}">
+{{ if eq .Params.mermaid true }}
+<script async src="https://unpkg.com/mermaid@8.5.0/dist/mermaid.min.js"></script>
```
Comment (sftim) [inline layouts/partials/css.html:23]: When this goes live, does it make sense to copy this into the built docs so it goes onto the Netlify CDN?

Comment (jimangel) [reply layouts/partials/css.html:23]: I like the idea of reducing external dependencies but I do want to make this very low-touch to support. Let me try a couple things...

**Comment by jimangel** [comment]
@sftim I've insourced the JS, commented where the origin is, and added a JavaScript error that looks like:

image

I also moved the JS implementation where we have other JS (head.html). Let me know if you think that would resolve your concerns. Thanks!

**Comment by jimangel** [comment]
> For me the ideal would be to have the website build invoke https://github.com/mermaid-js/mermaid.cli and then save the rendered output for the CDN to serve. That's likely to work with a very large audience of readers.

I completely agree that having the user compose an image would be better from a usability standpoint. I do think by having it as JavaScript enables community easy edits / updates without much effort. I started to look into Hugo web-hooks / markup approaches but couldn't find anything concrete to dive into.

That would be one alternative, if Hugo could generate the image on build based on a hook. Once again, my goal is to enable a feature without needing much support / effort to maintain.

**Comment by sftim** [comment]
@jimangel this looks like a good approach for now, with room to make things even better at some future point

**Comment by kbhawkey** [comment]
Hello @jimangel . I quickly looked at the mermaid docs and the content here.
A few thoughts about using this script/tool in the docs (I missed your intro).
The charts look nice. Do we need dynamic generation of charts? Could we use the tool offline to generate an image?
Is the mermaid code maintained?
Mermaid appears to have its own language to express the charts. Is this easy to use?
There seem to be some questions about security in the mermaid docs.
How do you change the fonts in the diagrams or size the generated chart on the page?
I have not used the Hugo render-image template, but it may be something to look into:
https://gohugo.io/getting-started/configuration-markup#image-markdown-example

**Comment by jimangel** [comment]
@kbhawkey before I address your questions, let me preface this by saying my primary goal was to introduce a gantt diagram for a releases page much like: https://en.wikipedia.org/wiki/Kubernetes#Release_Versions the general consensus was that it was a good idea.

Given the velocity of the releases (every quarter) my goal was to implement something similar to Wikipedia that could be changed every 12 weeks. With that, I wanted this feature to be easily modified and low effort to maintain.

More info about the releases page here: https://github.com/kubernetes/website/issues/20293

To be clear, we don't need to merge this PR and could still create an effective `releases` page but I think the chart is a good visual representation of support from the community.

> The charts look nice. Do we need dynamic generation of charts? Could we use the tool offline to generate an image?

I believe the velocity is the concern here. Dynamic generation really lowers the bar. The information should be additional not essential.

> Is the mermaid code maintained?

Looking at the source code https://github.com/mermaid-js/mermaid it appears it is actively maintained by ~140 contributors.

The code frequency appears to be healthy with the last commit ~3 days ago and the following chart indicates code velocity is also increasing:

image

Additionally, [hackmd.io](https://hackmd.io/) supports it too: https://hackmd.io/s/features#Mermaid

> Mermaid appears to have its own language to express the charts. Is this easy to use?

For my use case it seems very strait forward. I think the examples are a great reference.

> There seem to be some questions about security in the mermaid docs.

I'm not aware of security concerns, but that would definitely be worth addressing before merging this PR. Can you provide some additional context / links?

> How do you change the fonts in the diagrams or size the generated chart on the page?

The content can be modified via CSS and there's a few issues open regarding starting points for this. If you think this is a show stopper, I would be happy to include a custom style sheet with this PR.

> I have not used the Hugo render-image template, but it may be something to look into:
https://gohugo.io/getting-started/configuration-markup#image-markdown-example

I've seen this and would be 100% down to change this approach to render on build if it can be easily done. I haven't found a good example how we could trigger image builds via mermaid (or other charting software). Going back to my original comment, I want this to be easily changed and implementable without a high bar to contribute.

**Comment by zacharysarah** [comment]
@jimangel This is so plush. 🤩 

Does it make sense to add a link to mermaid syntax to `docs/en/contribute/style-guide/`? I'd be happy to take that on as a follow-up PR.

**Comment by jimangel** [comment]
+1 on the `good-first-issue` for the style-guide @zacharysarah. I would like feedback from @kbhawkey and if we all agree this PR will be good to approve / merge. Thanks!

**Comment by sftim** [comment]
/lgtm

Would also welcome feedback from @kbhawkey / @onlydole

**Comment by kbhawkey** [comment]
Hi. I am in favor of including well thought out and visually appealing diagrams in the docs 💹 .
My original question is why do we want _dynamic_ rendering of svg files?
I agree w/@sftim , ("For me the ideal would be to have the website build invoke https://github.com/mermaid-js/mermaid.cli and then save the rendered output for the CDN to serve ...").
One benefit of declaring the chart instructions in Markdown is that you can collaborate (in a pull request) on the chart design (and debug the graph instructions).
One downside of dynamic chart rendering is that the diagram/chart is not saved with the other site assets.
I think this could be a short term solution until something could be plugged into the docs build process. I wonder about an integration similar to PostCSS,
https://gohugo.io/hugo-pipes/postcss/#check-hugo-environment-from-postcssconfigjs

File: layouts/partials/head.html [hunk 2]
```diff
@@ -48,6 +48,11 @@
 <script src="{{ "js/bootstrap-4.3.1.min.js" | relURL }}"></script>
 <script src="{{ "js/sweetalert-2.1.2.min.js" | relURL }}"></script>
 
+{{ if eq .Params.mermaid true }}
+<!-- Copied from https://unpkg.com/mermaid@8.5.0/dist/mermaid.min.js -->
+<script async src="{{ "js/mermaid.min.js" | relURL }}"></script>
+{{ end }}
```
Comment (kbhawkey) [inline layouts/partials/head.html:54]: @jimangel , Question:
Do you know how the field `.Params.js` is used?

```
# config.toml
js = [
  "custom-jekyll/tags",
  "script"
]
```

Comment (jimangel) [reply layouts/partials/head.html:54]: I have no clue. It looks like it does the same thing twice (explicitly first then in shortcode after:
```
<script src="{{ "js/script.js" | relURL }}"></script>
<script src="{{ "js/custom-jekyll/tags.js" | relURL }}"></script>
{{ with .Params.js }}{{ range (split . ",") }}<script src="{{ (trim . " ") | relURL }}"></script><!-- custom js added -->
```

**Comment by BenTheElder** [comment]
trying to follow along, so this is still dynamically rendered right?

**Comment by jimangel** [comment]
Yep! However, there's a strong desire to eventually switch to generation on build; if possible.

**Pull request merged**